src/oge/helpers.py

```diff
@@ -287,6 +290,52 @@ def create_plant_ba_table(year: int) -> pd.DataFrame:
     return plant_ba
 
 
+def add_plant_operating_and_retirement_dates(df: pd.DataFrame) -> pd.DataFrame:
+    """Adds the operating and retirement dates of a plant to input data frame. The
+    operating date of a plant is taken as the earliest date among all generators'
+    operating date over all report dates. Likewise, the retirement date of a plant is
+    taken as the latest date among all generators' retirement date over all report
+    dates.
+
+    Note that the operating date is the date the generator began commercial operation.
+    The retirement date is the date of the scheduled or effected retirement of the
+    generator.
+
+    Args:
+        df (pd.DataFrame): table with a 'plant_id_eia' column.
+
+    Returns:
+        pd.DataFrame: original data frame with additional 'operating_date' and
+            'retirement_date' column.
+    """
+    generators_dates = load_data.load_pudl_table(
+        "denorm_generators_eia",
```

Comment:
this table will contain values for each year reported, so before we run our min and max operations, we need to drop duplicates. Before we drop duplicates though, we may need to do a groupby([plant_id, generator_id]).ffill() and .bfill() to make sure that we have complete values for all years

Reply:
We may also want to filter to only include data up to the latest_validated_year

Reply:
This is something I was working on in GRETA, but a similar pattern may work here:
```
min_operating = oge.load_data.load_pudl_table(
    "generators_eia860",
    year=earliest_data_year,
    end_year=latest_validated_year,
    columns=[
        "report_date",
        "plant_id_eia",
        "generator_id",
        "minimum_load_mw",
        "capacity_mw",
        "summer_capacity_mw",
        "winter_capacity_mw",
    ],
).sort_values(by=["plant_id_eia","generator_id","report_date"], ascending=True)

# fill missing capacity values
capacity_columns = ["minimum_load_mw", "capacity_mw", "summer_capacity_mw", "winter_capacity_mw"]

for col in capacity_columns:
    min_operating[col] = min_operating.groupby(["plant_id_eia","generator_id"])[col].bfill()
    min_operating[col] = min_operating.groupby(["plant_id_eia","generator_id"])[col].ffill()

# keep only the most recent year of data
min_operating = min_operating.drop_duplicates(subset=["plant_id_eia","generator_id"], keep="last")
```

Reply:
Thanks. Implemented.